File: usm-api/src/main/java/com/github/zly2006/usm/api/event/EventSource.java
package com.github.zly2006.usm.api.event;

import com.github.zly2006.usm.api.plugin.Plugin;

import java.util.List;

/**
 * Represents a source of events.
 * Maybe a plugin, a user, or a server.
 */
public interface EventSource {
    List<Listener> getListeners();
    void addListener(Plugin plugin, Listener listener);
    void removeListener(Plugin plugin, Listener listener);
    void callEvent(Event<?> event);
}

File: usm/src/main/kotlin/com/github/zly2006/usm/impl/EventSource.kt
package com.github.zly2006.usm.impl

import com.github.zly2006.usm.api.event.Event
import com.github.zly2006.usm.api.event.Listener
import com.github.zly2006.usm.api.plugin.Plugin

open class EventSource(

) : com.github.zly2006.usm.api.event.EventSource {
    val _listeners = mutableListOf<ListenerWrapper>()
    override fun getListeners(): MutableList<Listener> {
        TODO("Not yet implemented")
    }

    override fun addListener(plugin: Plugin?, listener: Listener?) {
        TODO("Not yet implemented")
    }

    override fun removeListener(plugin: Plugin?, listener: Listener?) {
        TODO("Not yet implemented")
    }

    override fun callEvent(event: Event<*>) {
        _listeners.forEach {
            it.callbackMap[event.javaClass]?.invoke(event)
        }
    }
}
